packages/walletconnect-v2/src/index.ts

```diff
@@ -86,27 +111,47 @@ export class WalletConnect extends Connector {
     this.events.emit(URI_AVAILABLE, uri)
   }
 
+  private async initializeProvider(
+    desiredChainId: number | undefined = this.defaultChainId
+  ): Promise<WalletConnectProvider> {
+    const ethProviderModule = await import('@walletconnect/ethereum-provider')
+    const rpcMap = this.rpcMap ? await getBestUrlMap(this.rpcMap, this.timeout) : undefined
+    const chainProps = this.getChainProps(this.chains, this.optionalChains, desiredChainId)
+
+    this.provider = await ethProviderModule.default.init({
+      ...this.options,
+      ...chainProps,
+      rpcMap,
+    })
+
+    return this.handleProviderEvents(this.provider)
+  }
+
+  private getChainProps(
+    chains: number[] | ArrayOneOrMore<number> | undefined,
+    optionalChains: number[] | ArrayOneOrMore<number> | undefined,
+    desiredChainId: number | undefined = this.defaultChainId
+  ): ChainsProps {
+    // Reorder chains and optionalChains if necessary
+    const orderedChains = getChainsWithDefault(chains, desiredChainId)
+    const orderedOptionalChains = getChainsWithDefault(optionalChains, desiredChainId)
+
+    // Validate and return the result
+    if (isArrayOneOrMore(orderedChains)) {
+      return { chains: orderedChains, optionalChains: orderedOptionalChains }
+    } else if (isArrayOneOrMore(orderedOptionalChains)) {
+      return { optionalChains: orderedOptionalChains, chains: orderedChains }
+    }
+
+    throw new Error('Either chains or optionalChains must have at least one item.')
+  }
+
   private isomorphicInitialize(
     desiredChainId: number | undefined = this.defaultChainId
   ): Promise<WalletConnectProvider> {
     if (this.eagerConnection) return this.eagerConnection
-
-    const rpcMap = this.rpcMap ? getBestUrlMap(this.rpcMap, this.timeout) : undefined
-    const chains = desiredChainId ? getChainsWithDefault(this.chains, desiredChainId) : this.chains
-
-    return (this.eagerConnection = import('@walletconnect/ethereum-provider').then(async (ethProviderModule) => {
-      const provider = (this.provider = await ethProviderModule.default.init({
-        ...this.options,
-        chains,
-        rpcMap: await rpcMap,
-      }))
-
-      return provider
-        .on('disconnect', this.disconnectListener)
-        .on('chainChanged', this.chainChangedListener)
-        .on('accountsChanged', this.accountsChangedListener)
-        .on('display_uri', this.URIListener)
-    }))
+    this.eagerConnection = this.initializeProvider(desiredChainId)
+    return this.eagerConnection
```

Comment:
maintain the existing codestyle
```suggestion
    return (this.eagerConnection = this.initializeProvider(desiredChainId))
```

Reply:
ew I hate this pattern so much